sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -269,220 +271,114 @@ Command *Scheduler::GraphBuilder::addCopyBack(Requirement *Req) {
-    Q2: | Write/Unmap 1 |<----+
-        +---------------+     |
-                              |-----This is blocked by user event
-        +---------------+     |
-    Q3: | Write/Unmap 2 |<----+
-        +---------------+
-    */
-
-    std::unique_ptr<UnMapMemObject> UnMapCmdUniquePtr(new UnMapMemObject(
-        *SrcReq, SrcAllocaCmd, Req, SrcQueue, /*UseExclusiveQueue*/ true));
-
-    if (!MapCmdUniquePtr || !UnMapCmdUniquePtr)
-      throw runtime_error("Out of host memory");
-
-    MapMemObject *MapCmd = MapCmdUniquePtr.release();
+  OpType Type = OpType::COPY;
+
+  if (SrcQueue->is_host())
+    Type = UPDATE;
+  else if (SrcQueue->get_device().is_cpu() &&
+           Req->MSYCLMemObj->getType() ==
+               detail::SYCLMemObjI::MemObjType::BUFFER && false)
+    Type = OpType::MAP_UNMAP;
+  else
+    Type = OpType::COPY;
+
+  Command *RetCmd = nullptr;
+
+  switch(Type) {
+  case OpType::MAP_UNMAP: {
+    MapMemObject *MapCmd = new MapMemObject(*SrcAllocaCmd->getAllocationReq(),
+                                            SrcAllocaCmd, Req, SrcQueue);
+
+    UnMapMemObject *UnMapCmd = new UnMapMemObject(
+        *SrcAllocaCmd->getAllocationReq(), SrcAllocaCmd, Req, SrcQueue);
+
+    UnMapCmd->MBlocking = true;
+    UnMapCmd->MCanEnqueue = false;
+
+    std::set<Command *> Deps = findDepsForReq(Record, Req, HostQueue);
     for (Command *Dep : Deps) {
       MapCmd->addDep(DepDesc{Dep, &MapCmd->MDstReq, SrcAllocaCmd});
       Dep->addUser(MapCmd);
     }
 
-    Command *UnMapCmd = UnMapCmdUniquePtr.release();
     UnMapCmd->addDep(DepDesc{MapCmd, &MapCmd->MDstReq, SrcAllocaCmd});
     MapCmd->addUser(UnMapCmd);
 
     UpdateLeafs(Deps, Record, Req->MAccessMode);
     AddNodeToLeafs(Record, UnMapCmd, Req->MAccessMode);
 
-    UnMapCmd->addDep(Req->BlockingEvent);
+    Req->MBlockingCommand = UnMapCmd;
 
-    RetEvent = MapCmd->getEvent();
-    if (MPrintOptionsArray[AfterAddHostAcc])
-      printGraphAsDot("after_addHostAccessor");
-    return UnMapCmd;
+    RetCmd = MapCmd;
+    break;
   }
+  case OpType::COPY:{
+    insertMemCpyCmd(Record, Req, HostQueue);
 
-  // In other cases insert two mem copy operations.
-  MemCpyCommand *DevToHostCmd = insertMemCpyCmd(Record, Req, HostQueue);
-  DevToHostCmd->setAccessorToUpdate(Req);
-  Command *HostToDevCmd =
-      insertMemCpyCmd(Record, Req, SrcQueue, /*UseExclusiveQueue*/ true);
-  HostToDevCmd->addDep(Req->BlockingEvent);
+    Command *UpdateHostAccCmd = insertUpdateHostReqCmd(Record, Req, HostQueue);
+
+    EmptyCommand *EmptyCmd = new EmptyCommand(HostQueue, Req);
+    EmptyCmd->addDep(DepDesc{UpdateHostAccCmd, EmptyCmd->getStoredRequirement(),
+                             SrcAllocaCmd});
+    UpdateHostAccCmd->addUser(EmptyCmd);
+
+    EmptyCmd->MBlocking = true;
+    EmptyCmd->MCanEnqueue = false;
+
+    UpdateLeafs({UpdateHostAccCmd}, Record, Req->MAccessMode);
+    AddNodeToLeafs(Record, EmptyCmd, Req->MAccessMode);
+
+    Req->MBlockingCommand = EmptyCmd;
```

Comment:
Can these lines be extracted in function like `insertEmptyCommand`?

Reply:
Probably, will try to add helper function.

Reply:
Didn't find a good solution for helper function. Will refactor the code to get rid of the duplication.